### PR TITLE
Task/bws81 delete user

### PR DESCRIPTION
**Pull request by Oleg-feriss**

### CODE REVIEW

File: backend/src/data/repositories/instance/instance.repository.ts [hunk 1]
```diff
@@ -41,6 +41,19 @@ class Instance {
...
+      .query()
+      .select()
+      .where({ 'createdBy': workerId });
+
+    if (!instances) {
+      return [];
+    }
+
+    return instances.map((instance) => instance.id);
+  }
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/instance/instance.repository.ts]:
```suggestion
  public async getByWorkerId(workerId: string): Promise<InstanceEntity[]> {
    const instances = await this.#InstanceModel
      .query()
      .select()
      .where({ 'createdBy': workerId });

  // .query() should already return array. You do no need to check on nullability

    return instances.map(Instance.modelToEntity);
  }
```

File: backend/src/data/repositories/slc-function/slc-function.repository.ts [hunk 2]
```diff
@@ -90,6 +90,19 @@ class SLCFunction {
...
+      .query()
+      .select()
+      .where({ 'createdBy': workerId });
+
+    if (!slcFunctions) {
+      return [];
+    }
+
+    return slcFunctions.map((slcFunction) => slcFunction.id);
+  }
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/slc-function/slc-function.repository.ts]:
```suggestion
  public async getByWorkerId(workerId: string): Promise<SLCFunctionEntity[]> {
    const slcFunctions = await this.#SLCFunctionModel
      .query()
      .select()
      .where({ 'createdBy': workerId });

    if (!slcFunctions) {
      return [];
    }

    return SLCFunction.modelToEntity(slcFunction as SLCFunctionM);
  }
```

File: backend/src/data/repositories/space/space.repository.ts [hunk 3]
```diff
@@ -45,6 +45,19 @@ class Space {
...
+      .query()
+      .select()
+      .where({ 'createdBy': workerId });
+
+    if (!space) {
+      return [];
+    }
+
+    return space.map((item) => item.id);
+  }
```
Comment (Yaroslaww-1) [inline backend/src/data/repositories/space/space.repository.ts]: Same as with slc functions and instances

File: backend/src/services/worker/worker.service.ts [hunk 4]
```diff
@@ -196,6 +226,31 @@ class Worker {
...
+      });
+    }
+
+    const instances = await this.#instanceRepository.getByWorkerId(id);
+    const space = await this.#spaceRepository.getByWorkerId(id);
+    const slcFunctions = await this.#slcFunctionRepository.getByWorkerId(id);
+
+    for (const id of instances) {
+      await this.#instanceService.delete(id);
+    }
```
Comment (Yaroslaww-1) [inline backend/src/services/worker/worker.service.ts]:
```suggestion
   await Promise.all(instances.map(instance => this.#instanceService.delete(instance .id));
```

Comment (Yaroslaww-1) [reply backend/src/services/worker/worker.service.ts]: `for await` processes requests sequentially. `Promise all` processes concurrently. 
You can find more about differences [here](https://www.linkedin.com/pulse/await-vs-promiseall-mohamed-kamel/)

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: backend/src/services/worker/worker.service.ts (hunk 4 above)
Comment (what1s1ove) [reply backend/src/services/worker/worker.service.ts]: agreed. prefer promise.all for all

File: backend/src/data/repositories/space/space.repository.ts (hunk 3 above)
Comment (what1s1ove) [reply backend/src/data/repositories/space/space.repository.ts]: +

File: backend/src/data/repositories/slc-function/slc-function.repository.ts (hunk 2 above)
Comment (what1s1ove) [reply backend/src/data/repositories/slc-function/slc-function.repository.ts]: +

File: backend/src/data/repositories/instance/instance.repository.ts (hunk 1 above)
Comment (what1s1ove) [reply backend/src/data/repositories/instance/instance.repository.ts]: +

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — DISMISSED** [review]

File: backend/src/services/worker/worker.service.ts [hunk 5]
```diff
@@ -196,6 +226,31 @@ class Worker {
...
+
+    const instances = await this.#instanceRepository.getByWorkerId(id);
+    const space = await this.#spaceRepository.getByWorkerId(id);
+    const slcFunctions = await this.#slcFunctionRepository.getByWorkerId(id);
+
+    await Promise.all(
+      instances.map((instance) => this.#instanceService.delete(instance.id)),
+    );
+    await Promise.all(
+      space.map((item) => this.#instanceService.delete(item.id)),
```
Comment (Yaroslaww-1) [inline backend/src/services/worker/worker.service.ts]: spaceService

File: backend/src/services/worker/worker.service.ts [hunk 6]
```diff
@@ -196,6 +226,31 @@ class Worker {
...
+    );
+    await Promise.all(
+      slcFunctions.map((slcFunction) =>
+        this.#instanceService.delete(slcFunction.id),
```
Comment (Yaroslaww-1) [inline backend/src/services/worker/worker.service.ts]: slcFunctionService

**Review by Yaroslaww-1 — CHANGES_REQUESTED** [review]

File: backend/src/services/worker/worker.service.ts [hunk 7]
```diff
@@ -184,6 +214,35 @@ class Worker {
...
+
+    const instances = await this.#instanceRepository.getByWorkerId(id);
+    const space = await this.#spaceRepository.getByWorkerId(id);
+    const slcFunctions = await this.#slcFunctionRepository.getByWorkerId(id);
+
+    await Promise.all(
+      instances.map((instance) => this.#instanceService.delete(instance.id)),
+    );
+    await Promise.all(
+      space.map((item) => {
```
Comment (junojei) [inline backend/src/services/worker/worker.service.ts]: B6
```suggestion
    const spaces = await this.#spaceRepository.getByWorkerId(id);
    const slcFunctions = await this.#slcFunctionRepository.getByWorkerId(id);

    await Promise.all(
      instances.map((instance) => this.#instanceService.delete(instance.id)),
    );
    await Promise.all(
      spaces.map((item) => {
```

File: backend/src/services/worker/worker.service.ts [hunk 8]
```diff
@@ -184,6 +214,35 @@ class Worker {
...
     }
 
+    const master = await this.#masterService.getMasterById(id);
+
+    if (master) {
+      throw new EAMError({
+        status: HttpCode.DENIED,
+        message: ExceptionMessage.MASTER_DELETE,
+      });
+    }
```
Comment (junojei) [inline backend/src/services/worker/worker.service.ts:224]: Why do we need this check?
Maybe add check for if this worker already deleted?

File: backend/src/services/worker/worker.service.ts (hunk 7 above)
Comment (what1s1ove) [reply backend/src/services/worker/worker.service.ts]: agreed!

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/services/worker/worker.service.ts (hunk 8 above)
Comment (Oleg-feriss) [reply backend/src/services/worker/worker.service.ts:224]: Master can not be deleted, this DELETE request is only for workers
image

**Review by what1s1ove — APPROVED** [review]

**Review by Yaroslaww-1 — APPROVED** [review]

**Pull request merged**